cardano-node/src/Cardano/Node/Protocol/Shelley.hs

```diff
@@ -91,37 +95,30 @@ mkConsensusProtocolShelley NodeShelleyProtocolConfiguration {
                             npcShelleyMaxSupportedProtocolVersion
                           }
                           files = do
-    genesis <- readGenesis npcShelleyGenesisFile
+    (genesis, genesisHash) <- readGenesis npcShelleyGenesisFile
     optionalLeaderCredentials <- readLeaderCredentials files
 
     return $
       Consensus.ProtocolRealTPraos
         genesis
-        initialNonce
+        (Nonce (Crypto.castHash genesisHash))
         (ProtVer npcShelleySupportedProtocolVersionMajor
                  npcShelleySupportedProtocolVersionMinor)
         npcShelleyMaxSupportedProtocolVersion
         optionalLeaderCredentials
-  where
-    -- The initial nonce, typically derived from the hash of Genesis config
-    -- JSON file.
-    --
-    -- WARNING: chains using different values of this parameter will be
-    -- mutually incompatible.
-    --
-    -- TODO: This should also be replaced with the hash of the Shelley genesis
-    -- config JSON file, which should be taken as an argument/configuration
-    -- parameter. But be careful, all testnets so far have used NeutralNonce,
-    -- changing the value will require starting a new testnet.
-    initialNonce = NeutralNonce
-
-readGenesis :: GenesisFile
+
+readGenesis :: Crypto.HashAlgorithm h
+            => GenesisFile
             -> ExceptT ShelleyProtocolInstantiationError IO
-                       (ShelleyGenesis TPraosStandardCrypto)
-readGenesis (GenesisFile file) =
-    firstExceptT (GenesisReadError file) $
-      ExceptT $ handle (\(e :: IOException) -> return $ Left $ show e) $
-        Aeson.eitherDecodeFileStrict' file
+                       (ShelleyGenesis TPraosStandardCrypto,
+                        Crypto.Hash h ByteString)
+readGenesis (GenesisFile file) = do
+    content <- handleIOExceptT (GenesisReadError file) $
+                 BS.readFile file
+    genesis <- firstExceptT (GenesisDecodeError file) $ hoistEither $
+                 Aeson.eitherDecodeStrict' content
+    let genesisHash = Crypto.hashRaw id content
```

Comment:
Isn't `hashRaw` [deprecated](https://github.com/input-output-hk/cardano-base/blob/7d795c3040ea7785812efa1c97864bbb41b15d3e/cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs#L269) in favour of `hashWith`?

Reply:
I'll use `hashWith` just as soon as we can update the dependencies! :-)

Reply:
> Fans of Cardano.Prelude might prefer...

Yeah, not me then :grin:

Reply:
Ah, of course, we need to update the dependency first :facepalm: